common/tools/dev-tool/README.md

```diff
@@ -0,0 +1,223 @@
+
+## Extending the Tool
+
+The source hierarchy matches the command hierarchy. Every sub-command has its own folder and `index.ts` file in `src/commands`, where `src/commands/index.ts` defines the behavior of the root `dev-tool` command, and each subfolder's `index.ts` file describes a nested sub-command. Every leaf node in the command tree ("leaf command") has its own TypeScript file. For example, `src/commands/about.ts` defines the behavior of the `dev-tool about` command, and `src/commands/package/resolve.ts` defines the behavior of the `dev-tool package resolve` command.
+
+### Command Interface
+
+Every command file's exports must implement the `CommandModule` interface defined in `src/util/commandModule.ts`. The interface requires that every command export a constant `commandInfo` that implements the `CommandInfo` interface defined in the same file. The `CommandInfo` interface specifies the name, description, and options (command-line arguments) of the command. The command module must also export an async handler function as its default export. Two helper functions, `leafCommand` and `subCommand` are provided to assist with development and to provide strong type-checking when
+extending dev-tool.
+
+### Creating a new leaf command
+
+To create a new leaf command in one of the existing sub-command, create a new TypeScript file for that command. Make sure that your module exports the required `commandInfo` and default handler function. When creating a command, use the `leafCommand` helper to get a strongly-typed `options` parameter for your handler.
+
+As an example, we can create a new `hello-world` command under the `dev-tool package` sub-command. The command will print out a string using the many different logging functions. It will accept an argument `--echo <string here>` that specifies the string to be printed.
+
+`src/commands/package/hello-world.ts`
+```typescript
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license
+
+import { createPrinter } from "../../util/printer";
+import { leafCommand } from "../../util/commandBuilder";
+
+const log = createPrinter("hello-world");
+
+export const commandInfo = {
+  name: "hello-world",
+  description:
+    "print a lovely message",
+  options: {
+      echo: {
+          kind: "string",
+          description: "override the message to be printed",
+          default: "Hello world!"
+      }
+  }
+} as const;
+
+export default leafCommand(commandInfo, async (options) => {
+    // Demonstrate the colorized command output.
+    log("Normal:", options.echo);
+    log.success("Success:", options.echo);
+    log.info("Info:", options.echo);
+    log.warn("Warn:", options.echo);
+    log.error("Error:", options.echo);
+    log.debug("Debug:", options.echo);
+
+    return true;
```

Comment:
In the value space of the JS runtime it is checked with `if (<return value>)` with no strict `true` check, so it will work with any truthy value. In type space the signature's return is `Promise<boolean>` so it will require a cast if you _really_ want to use a truthy/falsy non-boolean value.  The process will always `exit(1)` in cases where an exception was thrown or `false` was returned. Codes other than 1 generally aren't portable, but we could have a kind of internal convention. If a command really needs to exit in a specific way with a specific exit code that has a semantic meaning, it can just log an error and then call `process.exit(<code>)` to avoid all the error-handling code that would result from `return false;`.